early-attack-detector-py/src/utils.py

```diff
@@ -0,0 +1,119 @@
+from hexbytes import HexBytes
+import rlp
+import requests
+from web3 import Web3
+from time import time
+from concurrent.futures import ThreadPoolExecutor
+import functools
+import operator
+
+from src.constants import CONTRACT_SLOT_ANALYSIS_DEPTH, MASK, BOT_ID
+from src.logger import logger
+
+
+def calc_contract_address(w3, address, nonce) -> str:
+    """
+    this function calculates the contract address from sender/nonce
+    :return: contract address: str
+    """
+
+    address_bytes = bytes.fromhex(address[2:].lower())
+    return Web3.toChecksumAddress(Web3.keccak(rlp.encode([address_bytes, nonce]))[-20:])
+
+
+def is_contract(w3, address) -> bool:
+    """
+    this function determines whether address is a contract
+    :return: is_contract: bool
+    """
+    if address is None:
+        return True
+    code = w3.eth.get_code(Web3.toChecksumAddress(address))
+    return code != HexBytes("0x")
+
+def get_function_signatures(w3, opcodes) -> set:
+    """
+    this function will parse the opcodes and returns all the possible function signatures (essentially whenever we find a 4bytes in a PUSH4 instruction (e.g. PUSH4 0x42966c68))
+    """
+    function_signatures = set()
+    for i, opcode in enumerate(opcodes):
+        opcode_name = opcode.name
+        if opcode_name == "PUSH4" or opcode_name == "PUSH3":
+            function_signatures.add(f"0x{opcode.operand}")
+    return function_signatures
+
+
+def get_storage_addresses(w3, address) -> set:
+    """
+    this function returns the addresses that are references in the storage of a contract (first CONTRACT_SLOT_ANALYSIS_DEPTH slots)
+    :return: address_list: list (only returning contract addresses)
+    """
+    if address is None:
+        return set()
+    
+    tp_executor = ThreadPoolExecutor(max_workers=10)
+    checksumed_address = Web3.toChecksumAddress(address)
+    futures = [tp_executor.submit(w3.eth.get_storage_at, checksumed_address, i) for i in range(CONTRACT_SLOT_ANALYSIS_DEPTH)]
+    tp_executor.shutdown(wait=True)
+    futures_result = [f.result() for f in futures]
+    all_addresses = [[result[0:20].hex(), result[12:].hex()] for result in futures_result if result != HexBytes('0x0000000000000000000000000000000000000000000000000000000000000000')]
+    all_addresses = list(set(functools.reduce(operator.iconcat, all_addresses, [])))
```

Comment:
Should we be manually filtering out zero address from `all_addresses`? It will be frequent to store a zero address when there's a small value stored, e.g. `HexBytes('0x0000000000000000000000000000000000000000000000000000000000000352')` and we're storing its 20 left-most chars